src/main/java/address/sync/CloudSimulator.java

```diff
@@ -1,129 +1,670 @@
+    private RawCloudResponse getEmptyResponse(int responseCode) {
+        return new RawCloudResponse(responseCode, null, convertToInputStream(getStandardHeaders()));
+    }
+
+    private void resetQuotaAndRestartTimer() {
+        long nextResetTime = getNextResetTime();
+        rateLimitStatus.setQuotaResetTime(nextResetTime);
+        rateLimitStatus.setQuotaRemaining(API_QUOTA_PER_HOUR);
+        if (timer != null) {
+            timer.stop();
+        }
+        int timeout = (int) (rateLimitStatus.getQuotaReset() - LocalDateTime.now().toEpochSecond(getSystemTimezone()));
+        timer = new TickingTimer("Cloud Quota Reset Time", timeout, this::printTimeLeft,
+                                this::resetQuotaAndRestartTimer, TimeUnit.SECONDS);
+        timer.start();
+    }
+
+    private void printTimeLeft(int timeLeft) {
+        System.out.println(timeLeft + " seconds remaining to quota reset.");
+    }
+
+    private List<CloudPerson> filterPersonsByTime(List<CloudPerson> personList, LocalDateTime time) {
+        return personList.stream()
+                .filter(person -> !person.getLastUpdatedAt().isBefore(time))
+                .collect(Collectors.toList());
+    }
+
+    private long getNextResetTime() {
+        LocalDateTime curTime = LocalDateTime.now();
+        LocalDateTime nearestHour = LocalDateTime.of(
+                curTime.getYear(), curTime.getMonth(), curTime.getDayOfMonth(), curTime.getHour() + 1,
+                0, 0, 0);
+
+        return nearestHour.toEpochSecond(getSystemTimezone());
+    }
+
+    private ZoneOffset getSystemTimezone() {
+        LocalDateTime localDateTime = LocalDateTime.now();
+        ZonedDateTime zonedDateTime = localDateTime.atZone(ZoneOffset.systemDefault());
+        return zonedDateTime.getOffset();
+    }
+
+    private boolean shouldSimulateNetworkFailure() {
+        return shouldSimulateUnreliableNetwork && RANDOM_GENERATOR.nextDouble() <= FAILURE_PROBABILITY;
+    }
+
+    private boolean shouldSimulateSlowResponse() {
+        return shouldSimulateUnreliableNetwork && RANDOM_GENERATOR.nextDouble() <= FAILURE_PROBABILITY;
+    }
+
+    private RawCloudResponse getNetworkFailedResponse() {
+        System.out.println("Cloud simulator: failure occurred! Could not retrieve data");
+        return new RawCloudResponse(HttpURLConnection.HTTP_INTERNAL_ERROR, null, null);
+    }
+
+    private boolean isWithinQuota(int quotaUsed) {
+        System.out.println("Current quota left: " + rateLimitStatus.getQuotaRemaining() + ", using " + quotaUsed);
+        return quotaUsed <= rateLimitStatus.getQuotaRemaining();
+    }
+
+    private boolean isExistingPerson(List<CloudPerson> personList, CloudPerson targetPerson) {
+        return personList.stream()
+                .filter(person -> person.getFirstName().equals(targetPerson.getFirstName())
+                        && person.getLastName().equals(targetPerson.getLastName()))
+                .findAny()
+                .isPresent();
```

Comment:
I think it would be better if we use `personList.contains(targetPerson)` with `CloudPerson` having equality implemented. This would be helpful in the case, say, `CloudPerson` now is compared using their `generatedId`.